Card styling/UX improvements

brandonkelly opened the pull request:
### Description

Cards, chips, and inline-editable Matrix blocks are now styled more consistently:

- Each have a consistent background color (either gray or based on the entry type color), with a subtle 1px border around it, and subtle box shadow.
- Cards and inline-editable Matrix blocks have titlebars that have a slightly darker background color, which include the icon and entry type name (if applicable) on the left, and action items on the right.
- Card thumbnails are now styled more consistently with large chips, and are now right-aligned, so card content lines up regardless of whether there’s a thumbnail.

![Various element cards and chips](https://github.com/user-attachments/assets/83d0ec2b-12a7-416f-8f0f-815b5df40ab3)

A couple UX improvements are included as well:

- Nested element cards’ titles are no longer hyperlinked. Instead, an edit button is included in the toolbar, which opens the slideout editor. <kbd>Command</kbd>/<kbd>Ctrl</kbd>-clicking on an edit button opens the element’s edit page in a new window (if it has one).
- Card body text is now user-selectable.

### Related issues

- #15651
- #16632
- #16763
- #16185
- #16186
- #16670

brandonkelly committed:
Card styling/UX improvements

brandonkelly committed:
Style improvements

brandonkelly committed:
hyperlink=true by default for chips

Consistent with cards, and fixes a bug where element chips would lose their hyperlink on re-render

brandonkelly committed:
getCardTitlebarLabel() → getCardTitle()

brandonkelly committed:
Let chips in cards have normal padding

brandonkelly committed:
Fixed card previews

brandonkelly committed:
Merge branch '5.7' into feature/cms-1367-improve-card-stylingux

brandonkelly committed:
Don't hyperlink pasted element chips

iainhenderson commented:
Brilliant thanks @brandonkelly

i-just reviewed (changes requested):
Only a few of issues that I noticed:

- when you create a new nested entry in a matrix field (view mode cards) for an existing or brand new root entry, once the nested entry is saved, you can’t use the actions menu (3 dots); the edit icon triggers two edit slideouts, reordering doesn’t seem to work (via diamond icon); once you reload the page (or save the root entry), all of the above starts working as expected
- matrix in cards view mode (with the standard title field), when you duplicate a nested entry via actions menu > duplicate, it shows the title as hyperlinked; if you reload, the hyperlink is gone; same with copy and paste; this doesn’t happen when duplicating in the element index view mode;
- if you’ve chosen to show Entry Type in the card and you’re logged in as a non-admin, the Entry Type name is hyperlinked, but you don’t have permissions to edit Entry Types; maybe it would be best not to hyperlink it for non-admins?
- if we’re hyperlinking the Entry Type in the card, should we also hyperlink the Section (to link to the Section’s settings)?
- on the element index page in table view mode, there seems to be a large gap between the last item on the list and the footer

BTW, I love that you can click on the name of the entry type to edit it straight away (e.g. from the field edit page).

Once released, the CKEditor should be updated so that we don’t show the edit button and don’t hyperlink the title in the nested element cards.

brandonkelly committed:
Merge branch '5.7' into feature/cms-1367-improve-card-stylingux

brandonkelly committed:
Merge branch '5.7' into feature/cms-1367-improve-card-stylingux

brandonkelly committed:
Don't hyperlink newly-added cards

brandonkelly committed:
Don't link to components if the user isn't an admin

brandonkelly committed:
Don't create href-less <a> tags for elements w/o cp-url attributes

brandonkelly committed:
Render entry section attributes as chips

brandonkelly commented:
Thanks @i-just!

> * when you create a new nested entry in a matrix field (view mode cards) for an existing or brand new root entry, once the nested entry is saved, you can’t use the actions menu (3 dots); the edit icon triggers two edit slideouts, reordering doesn’t seem to work (via diamond icon); once you reload the page (or save the root entry), all of the above starts working as expected

Good catch. Looks like I introduced that in fd65d23b3312d7cce8e5b8670db66f7e154e3289 for #16893, so fixed on the 5.7 branch.

> * matrix in cards view mode (with the standard title field), when you duplicate a nested entry via actions menu > duplicate, it shows the title as hyperlinked; if you reload, the hyperlink is gone; same with copy and paste; this doesn’t happen when duplicating in the element index view mode;

Fixed!

> * if you’ve chosen to show Entry Type in the card and you’re logged in as a non-admin, the Entry Type name is hyperlinked, but you don’t have permissions to edit Entry Types; maybe it would be best not to hyperlink it for non-admins?

Doh, yeah. Updated all components’ `getCpEditUrl()` methods to return `null` for non-admins.

> * if we’re hyperlinking the Entry Type in the card, should we also hyperlink the Section (to link to the Section’s settings)?

Agreed. Updated `section` attributes to render as chips.

> * on the element index page in table view mode, there seems to be a large gap between the last item on the list and the footer

Are you referring to the 24px top margin on the footer? If so that’s intentional and hasn’t changed in 5.7 or this PR.

![The top margin of the control panel footer](https://github.com/user-attachments/assets/e66bada0-19f2-4f6e-b54d-7cd8cfc52b79)

> Once released, the CKEditor should be updated so that we don’t show the edit button and don’t hyperlink the title in the nested element cards.

Yep. Actually it would be safe to update now. The instance config passed to `app/render-elements` just needs to include:

```js
showEditButton: false,
hyperlink: false,
```

Those values will just be ignored if running < 5.7.

i-just commented:
Thanks!

Looks like the edit icon still opens up two slideouts (when you create a new nested entry in a matrix field (view mode cards) for an existing or brand new root entry).

> Are you referring to the 24px top margin on the footer? If so that’s intentional and hasn’t changed in 5.7 or this PR.

I meant the space between the last item and the footer. It looks about 15px taller on this PR vs 5.x branch. I measured the space between the bottom of the title and the start of the footer, and it’s around 30px on 5.x and 45px on this branch. (The `#footer` `margin-block` is 9px on 5.x and 5.7 and 24px on this branch.)

5.x
<img width="1616" alt="Screenshot 2025-03-31 at 13 34 28" src="https://github.com/user-attachments/assets/e3dc8c19-d7e1-42cd-9759-6d58837cac7f" />

this branch
<img width="1619" alt="Screenshot 2025-03-31 at 13 35 28" src="https://github.com/user-attachments/assets/7d3012ae-a3a4-41fb-9545-7597df2411ed" />


> Yep. Actually it would be safe to update now. The instance config passed to app/render-elements just needs to include

Yup, I already have it done locally :) I’ll PR it at some point today.

gcamacho079 commented on src/helpers/Cp.php:
```diff
@@ -688,39 +730,67 @@ public static function elementCardHtml(ElementInterface $element, array $config
         $html = Html::beginTag('div', $attributes) .
+            Html::beginTag('div', ['class' => 'card-titlebar']) .
+            Html::beginTag('div', [
+                'class' => ['flex', 'flex-nowrap', 'flex-gap-xs'],
```
```suggestion
                'class' => ['flex', 'flex-nowrap', 'flex-gap-s'],
```
The current spacing seems a little close to me; would this work?

brandonkelly replied:
Sure, done!

brianjhanson reviewed (approved):
For a bit I tried to figure out if we could make matrix blocks and cards share the same rendering path, but it looked like that was a bigger job than was worth taking on right now.

thupsi commented:
Hi! I've noticed some bugs with the new styles:

![image](https://github.com/user-attachments/assets/2bc75848-fc01-4189-bac2-05d5a1ce716d)

1. The active tab in the "Row" block has a white background, instead of a yellow tint.
2. The "Column" blocks, having no color set, are inheriting the yellow shade for their titlebar.
3. The text in those blocks also inherits the text color from the parent "Row" block, so it's not gray but dark brown/yellow instead (might not be immediately noticeable in the screenshot). 

In addition to the above I would like to make a proposal to **add "White" to the available colors for entry types**. Since the default now is implicitly "Gray", it would enable keeping the palette muted but have visual grouping and hierarchy by lightness as opposed to color, like in the example bellow:

![image](https://github.com/user-attachments/assets/497f39de-cf68-4e77-b78a-a450e4839f58)

brandonkelly committed:
Merge branch '5.7' into feature/cms-1367-improve-card-stylingux

brandonkelly committed:
Tweaks & fixes

brandonkelly commented:
> Looks like the edit icon still opens up two slideouts (when you create a new nested entry in a matrix field (view mode cards) for an existing or brand new root entry).

@i-just Doh, missed that. Fixed now!

> * The active tab in the "Row" block has a white background, instead of a yellow tint.
> * The "Column" blocks, having no color set, are inheriting the yellow shade for their titlebar.
> * The text in those blocks also inherits the text color from the parent "Row" block, so it's not gray but dark brown/yellow instead (might not be immediately noticeable in the screenshot).

@thupsi Thanks for testing! Fixed.

brandonkelly committed:
Release notes

[ci skip]

thupsi commented:
@brandonkelly I found another small CSS color bug. Text in CKEditor fields inherits the text color of the containing block:
![image](https://github.com/user-attachments/assets/25239ac1-8194-42a1-bcbf-c44da41a912c)

brandonkelly commented:
@thupsi Thanks, posted that on CKEditor’s repo: https://github.com/craftcms/ckeditor/issues/390

iamtompickering commented:
@brandonkelly not sure if it's just me being a bit picky (apologies if I am!), but the font sizing of Blocks as inline-editable blocks feels a little off now to me. I made a few minor CSS changes using a CP CSS plugin to make a few adjustments. 

Before: 
![CleanShot 2025-04-13 at 15 27 06@2x](https://github.com/user-attachments/assets/212e30c5-ba44-4f38-81c8-1e3f273402c2)

![CleanShot 2025-04-13 at 15 38 49@2x](https://github.com/user-attachments/assets/7eb93a4d-e273-450b-b3bb-add240a646f4)

After: 
![CleanShot 2025-04-13 at 15 43 52@2x](https://github.com/user-attachments/assets/186cb1cb-96eb-426a-a659-c45e2d3123fa)


![CleanShot 2025-04-13 at 15 39 53@2x](https://github.com/user-attachments/assets/153fbcc8-557c-4514-ab3d-3af3fe092816)


The sizing adjustments I made feel a little more in keeping with the spacing of everything else around it in my opinion. Just thought I'd share my opinions, hope that's okay!

brandonkelly commented:
@iamtompickering Thank for the feedback. I prefer the smaller font size for entry type labels, as a way to set it apart from the tab labels, but I agree the tightened-up inline-editable blocks looks a bit better. So just made that change for the next release. (ea36d2225a1d1d711f4d051be7335597cef6cb69)

iamtompickering commented:
@brandonkelly completely fair and understand the logic behind it! I'm sure I'll learn to love the smaller font size!
